docs/src/main/sphinx/develop/client-protocol.rst

```diff
@@ -0,0 +1,257 @@
+======================
+Trino client REST API
+======================
+
+The Trino client allows users to submit Trino queries and view results.  This
+note documents the REST API used by the Trino client.
+
+HTTP methods
+============
+
+* A ``POST`` to ``/v1/statement`` runs the query string in the ``POST`` body,
+  and returns a JSON document containing the query results.  If there are more
+  results, the JSON document will contain a ``nextUri`` URL attribute.
+* A ``GET`` to the ``nextUri`` attribute returns the next batch of query results.
+* A ``DELETE`` to ``nextUri`` terminates a running query.
+
+Overview of query processing
+============================
+
+A Trino client request is initiated by an HTTP ``POST`` to the endpoint
+``/v1/statement``, with a ``POST`` body consisting of the SQL query string.
+The caller may set header ``X-Trino-User`` to the username for the session,
+as well as a long list of other headers, documented below.
+
+If the client request returns an HTTP 503, that means the server was busy,
+and the client should try again in 50-100 milliseconds.  Any HTTP status other
+than 503 or 200 means that the query has failed.
+
+The ``/v1/statement`` ``POST`` request returns a JSON document of type
+``QueryResults``, as well as a collection of response headers.  The
+``QueryResults`` document will contain an ``error`` field of type
+``QueryError`` if the query has failed, and if that object is not present,
+the query succeeded.  Important members of ``QueryResults`` are documented
+below.
+
+If the ``data`` field of the JSON document is set, it contains a list of the
+rows of data, and the ``columns`` field will also be set to a list of the
+names and types of the columns returned by the query.  Most of the response
+headers should be treated like browser cookies by the client, and echoed back
+as request headers in subsequent client requests, as documented below.
+
+If the JSON document returned by the ``POST`` to ``/v1/statement`` does not
+contain a ``nextUri`` link, the query has completed, either successfully or
+unsuccessfully, and no additional requests need to be made.  If the
+``nextUri`` link is present in the document, there are more query results
+to be fetched.  The client should loop executing a ``GET`` request
+to the ``nextUri`` returned in the ``QueryResults`` response object until
+``nextUri`` is absent from the response.
```

Comment:
Is there some sort of timing requirement? Or how long does Trino stand by and is ready to provide more results before a query is abandoned? What if a query was abandoned and then you submit that request.. will it resubmit the query and get the correct sub-resultset / page .. or will it fail?

Reply:
I don't know.  I'll discuss with David P. and follow up in this PR.

Reply:
> Is there some sort of timing requirement? 

yes, and it's configurable. (`query.client.timeout` i think) 
That is not part of the protocol definition IMO. Every communication protocol has timeouts of some sorts (often: multiple different timeouts)

> What if a query was abandoned and then you submit that request

the request fails

Reply:
imho it would be good to point developers to more info about that somewhere but we can leave it out for now